platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/BiProvider.java

```diff
@@ -41,37 +41,42 @@ protected String toStringNoReentrance() {
 
     @Override
     public boolean calculatePresence(ValueConsumer consumer) {
-        if (!left.calculatePresence(consumer) || !right.calculatePresence(consumer)) {
-            return false;
+        try (EvaluationContext.ScopeContext ignored = openScope()) {
+            if (!left.calculatePresence(consumer) || !right.calculatePresence(consumer)) {
```

Comment:
❓ Shouldn't the context be passed down to both `calculatePresence` calls here?

Reply:
> ❓ Shouldn't the context be passed down to both `calculatePresence` calls here?

No, not in the current implementation. The context is being passed around (as a token) when we're calling methods of the provider being currently evaluated, to highlight two things:
- `this` shouldn't be added to the context again
- we can safely evaluate "inner" providers/user code inside the method.

I'm thinking about alternative designs where providers can indeed receive context and register themselves (saving us a ThreadLocal lookup), but these ideas don't tick all the boxes yet to ensure that all Provider implementations are forced to follow the protocol.

Reply:
I see.